File: 0003-longest-substring-without-repeating-characters.c
int lengthOfLongestSubstring(char* s) {
    unsigned short s_len = strlen(s);

    unsigned short longest_l = 0;
    unsigned short current_l = 0;
    
    short last_seen[96];

    for(unsigned short i = 0; i < 96; ++ i) 
        last_seen[i] = -1;

    for(unsigned short i = 0; i < s_len; ++ i) {
        unsigned short current_char = ((unsigned short) s[i] - 32);

        if(last_seen[current_char] < (i - current_l)) {
            
            last_seen[current_char] = i;
            ++ current_l;

            if(current_l > longest_l) {
                longest_l = current_l;
            }
        } else {
            current_l = i - last_seen[current_char];
            last_seen[current_char] = i;
        }
    }

    return longest_l;
}

File: dev-rust/0003-longest-substring-without-repeating-characters.rs
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut s_len = s.len();

        let mut longest_l: i16 = 0;
        let mut current_l: i16 = 0;

        let mut last_seen: [i16; 96] = [-1; 96];
            
        for (_i, _current_char) in s.chars().enumerate() {
            let i = (_i as i16);
            let current_char = (_current_char as i16) - 32;

            if last_seen[current_char as usize] < (i - current_l) {
                last_seen[current_char as usize] = i;
                current_l += 1;

                if current_l > longest_l {
                    longest_l = current_l;
                }
            } else {
                current_l = i - last_seen[current_char as usize];
                last_seen[current_char as usize] = i;
            }
            
        }

        return longest_l as i32;
    }
}
